### PR TITLE
[linalg] Add quantized version of `conv_3d_ncdhw_fcdhw`

### PR DESCRIPTION
**Pull request by ubfx**
This patch adds the quantized 3d convolution operator `conv_3d_ncdhw_fcdhw_q`. This is the "channel-first" dimension ordering used by PyTorch and others.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
I dont know where these uses are coming from, but I'd like to just say that addition of new convolution ops is kind of getting out-of-hand. We have a lot and it doesnt cover all use cases, but it seems like might be time to move convolutions out of ODS generated named ops.

@javedabsar  something you are interested in?

**Comment by ubfx** [comment]
> I dont know where these uses are coming from

I can't speak for the other Ops, but the "channel first" ones I've been adding are in order to support lowering torch IR to linalg without additional transpositions in torch-mlir.

**Comment by rengolin** [comment]
> I dont know where these uses are coming from, but I'd like to just say that addition of new convolution ops is kind of getting out-of-hand. We have a lot and it doesnt cover all use cases, but it seems like might be time to move convolutions out of ODS generated named ops.

I second that. While we're not directly interested in convolution ops (so won't have time to move them to ODS ourselves), I strongly suggest those interested do so instead of adding yet-another op in the large soup of slightly different semantics.

There has to be a way to common them all up using ODS + C++ code that OpDSL just couldn't. This is the reason we moved `matmul` to ODS and we're much better for it.

For the record, the medium-term plan is to move all relevant Linalg operations to ODS and remove OpDSL for good, the code merged in this PR will not live long. So if you care about these operations, please come up with a reasonable (parametrized) implementation in ODS and move the rest there, too.

**Comment by MaheshRavishankar** [comment]
> > I dont know where these uses are coming from, but I'd like to just say that addition of new convolution ops is kind of getting out-of-hand. We have a lot and it doesnt cover all use cases, but it seems like might be time to move convolutions out of ODS generated named ops.
> 
> I second that. While we're not directly interested in convolution ops (so won't have time to move them to ODS ourselves), I strongly suggest those interested do so instead of adding yet-another op in the large soup of slightly different semantics.
> 
> There has to be a way to common them all up using ODS + C++ code that OpDSL just couldn't. This is the reason we moved `matmul` to ODS and we're much better for it.
> 
> For the record, the medium-term plan is to move all relevant Linalg operations to ODS and remove OpDSL for good, the code merged in this PR will not live long. So if you care about these operations, please come up with a reasonable (parametrized) implementation in ODS and move the rest there, too.

Yeah, that said. I really cannot in good faith put this on people trying to connect it to Torch. This is really just broken in Linalg. We need to fix this "properly", and that will take time, but in the meantime I think we need to provide some way forward. Not sure how to balance that without adding to the problem. @ubfx  is there a way around adding a new operator?

**Comment by banach-space** [comment]
> I can't speak for the other Ops, but the "channel first" ones I've been adding are in order to support lowering torch IR to linalg without additional transpositions in torch-mlir.

Avoiding unnecessary transposition is good, but will this convolution get vectorized? If not, does this make any difference? Without vectorization, things will be slow _with_ and _without_ transposing.

Just wanted to draw people's attention to the fact that adding a new named Op, especially a convolution, is just one step towards good lowering 😅 And  that there are some non-trivial gaps in how convolutions are vectorized.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just waiting for a few more eyes on this before we land this.

**Comment by ubfx** [comment]
> @ubfx is there a way around adding a new operator?

We can of course live with the transposition arrangement for a while, particularly the 3d convolution is probably less important. But we also have 2d depthwise convolution which currently has no quantized channel first named Op available, and which we would ideally like to have a named representation of.

Just to be clear - I would also prefer having a single `quantized_conv3d` Op (or similar), which captures the whole subset of dimension orderings. It's just that the `linalg` convolution Ops are probably among the Ops with the most downstream uses, so such an op would likely need much more careful and elaborate design than I could provide. Of course - if such an op or concept existed, I'd be happy to port our uses in torch-mlir accordingly.


> Avoiding unnecessary transposition is good, but will this convolution get vectorized? If not, does this make any difference? Without vectorization, things will be slow _with_ and _without_ transposing.

I'm not sure if/how well vectorization is performed on the CPU backend. Personally, I use linalg graphs lowered from torch IR as a starting point for compilation to a custom dataflow accelerator. Graph partitioning becomes increasingly difficult when having to consider additional transposition ops at the inputs and outputs of the conv ops. In quantized graphs, DQ/Q ops might also sneak in between the convolution and transpositions, which makes pattern matching even more annoying and less reliable. And if a un-fused transposition actually makes it into the final partitioned graph, that's a huge cost on my hardware.
I'm not trying to hijack `linalg` for my specific use case - but it does feel like the abstract goal of "avoiding unnecessary transpositions" is a general good, even though some backends and lowering paths might only benefit minimally.

**Comment by rengolin** [comment]
> It's just that the `linalg` convolution Ops are probably among the Ops with the most downstream uses, so such an op would likely need much more careful and elaborate design than I could provide.

This doesn't scale. If no downstream user is "able to provide" a reasonable implementation upstream and just start dumping all variations of the same op to suit their needs, the dialect will be bloated and unusable.

**Comment by rengolin** [comment]
> but it does feel like the abstract goal of "avoiding unnecessary transpositions" is a general good, even though some backends and lowering paths might only benefit minimally.

This is a false correlation. Yes, avoiding unnecessary transposition is a good thing, but the way you're achieving is by bloating the namespace beyond usefulness. This is the same problem we have with `matmul_transpose_a` and `matmul_transpose_b` etc. It does not actually _solve_ the problem, just moves it somewhere else.

See https://github.com/llvm/llvm-project/pull/104783 for a path to a proper solution.

**Comment by banach-space** [comment]
I share Renato's concerns and agree with his perspective.

As I understand, we're aligned on the need to reduce the number of named operations in Linalg, and there’s already ongoing work in this direction. This PR, while aiming to support a specific downstream target, seems to move us away from that goal and, therefore, feels counterproductive.

Supporting downstream targets is certainly valuable, and I recognize the benefit of this approach for your project:

> I use Linalg graphs lowered from Torch IR as a starting point for compilation to a custom dataflow accelerator.

However, from a pragmatic and long-term perspective, we should carefully consider the potential consequences. The risk is that downstream targets might get discontinued, or future downstream projects may have different needs. This could lead to even more named Ops, which would be challenging to maintain and reconcile.

I appreciate the work you've put into this PR, and I don’t want to discourage contributions, but I believe it would be beneficial to explore alternative approaches that align with our broader goals.

**Comment by MaheshRavishankar** [comment]
I think we need to make a decision here.
1) We freeze any additions to Linalg named ops.
2) We allow uses of named ops of this form to come in and then build a "generalized convolution op" in Linalg that accounts for all convolution names ops.

I am not really comfortable with 1 cause we dont have a plan to fix it for downstream users. Without a solid plan and some one signed up to work on it, I would rather go with (2). I know I blocked this PR to kick start a discussion, but that isnt really fair to the contributor here (it is on us as defacto maintainers to provide a better solution).
If we are on-board with (2), we can let this one go in and add it to the list of "ops we need to deprecate".

**Comment by stellaraccident** [comment]
Couple of comments: 

1. I think the most widely used op was probably the matmuls, and we just got that reconverged (still some cleanup but the main thrust resolved).
2. Convolution naturally has more variants and practical decisions need to be made. I assume the answer is not one op. But neither is it ten.
3. It is a historic anomaly that channels last got the priority when the most dominant framework operates channels first. We can't block some amount of progress on fixing that bug, regardless of POR. It would be good to treat these as high cost exceptions and only do it with a strong expressed need, though... Since it adds to the work needed to fix the bigger issue.
4. We should publish a plan of record on where convolution needs to get to, informed by the matmul work. It's unfair to just tell people no on such an important case without so much as a written plan on what to do about it.

**Comment by stellaraccident** [comment]
We joined this PR with the other conv PR and moved discussion to the forum: https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863

File: mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml
Comment (javedabsar1) [inline mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml:4033]: s/is performed on the operands to the inner multiply/
  is performed on the operands of the inner multiply ?

File: mlir/test/Dialect/Linalg/roundtrip.mlir
Comment (javedabsar1) [inline mlir/test/Dialect/Linalg/roundtrip.mlir:704]: Do both (%image, %filter) have to be same type?

**Pull request closed without merging**
